app/controllers/application_controller.rb

```diff
@@ -5,6 +5,174 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    #added below
+    enable :sessions
+    set :session_secret, "hashtagwars"
   end
 
-end
\ No newline at end of file
+  get '/' do 
+    erb :index
+  end
+
+  get '/login' do 
+    if is_logged_in?
+      redirect '/tweets'
+    else
+      erb :"users/login"
+    end
+  end
+
+  post '/login' do
+    @user = User.find_by_username(params[:username])
+    session[:id] = @user[:id]
+    if @user.password == params[:password]
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/signup' do
+    if is_logged_in?
+      redirect '/tweets'
+    else
+      erb :"users/create_user"
+    end
+  end
+
+  post '/signup' do
+    if params[:username] == "" || params[:email] == "" || params[:password] == ""
+      redirect '/signup'
+    end
+    @user = User.find_by(username: params[:username])
+    if @user == nil
+      @user = User.create(username: params[:username], email: params[:email], password: params[:password])
+      @user.save
+
+      session[:id] = @user.id
+
+      redirect '/tweets'
+    elsif @user.password == params[:password]
+      session[:id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets' do
+    if is_logged_in?
+      @user = current_user
+      @tweets = Tweet.all
+      erb :"tweets/tweets"
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/logout' do
+    if is_logged_in?
+      session.clear
+      redirect '/login'
+    else
+      redirect '/'
+    end
+  end
+
+  get '/tweets/new' do
+    if is_logged_in?
+      erb :"tweets/create_tweet"
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets/new' do
+    if params[:content] == ""
+      redirect '/tweets/new'
+    else
+      new_tweet = Tweet.create(content: params[:content], user_id: session[:id])
+      new_tweet.save
+      redirect '/tweets'
+    end
+  end
+
+  get '/tweets/:id' do
+    @tweet = Tweet.find_by_id(params[:id])
+
+    if is_logged_in?
+      erb :"tweets/show_tweet"
+    else
+      redirect '/login'
+    end
+
+  end
+
+  get '/tweets/:id/edit' do
+
+    if session[:id] == nil
+      redirect '/login'
+    end
+    @tweet = Tweet.find_by_id(params[:id])
+    if session[:id] == @tweet.user_id
+
+      erb :"tweets/edit_tweet"
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post '/tweets/:id/edit' do 
+
+    if session[:id] == nil
+      redirect '/login'
+    end
+    if params[:content] == ""
+      redirect "/tweets/#{params[:id]}/edit"
+    end 
+    @tweet = Tweet.find_by_id(params[:id])
+    if session[:id] == @tweet.user_id
+      @tweet.content = params[:content]
+      @tweet.save
+      redirect "/tweets/#{params[:id]}"
+    end
+
+  end
+
+  post '/tweets/:id/delete' do
+
+    @tweet = Tweet.find_by_id(params[:id]) #finds correct tweet
+    if session[:id] == @tweet.user_id 
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the users tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully simply returning nil if the user is trying to delete a tweet that they do not have access to.